Deprecate OCSP_xxx API for OSSL_HTTP_xxx

richsalz opened the pull request:
This is the replacement for #13671. It deprecated various ```OCSP_``` functions for their ```OSSL_HTTP_``` replacements.

I did not add documentation; things that were undocumented before, remain undocumented.

I also noticed that, arguably, some functions are missing; e.g., while there's ```OSSL_HTTP_REQ_CTX_free```, and there's ```OCSP_sendreq_new``` which returns a context that free uses, there's no generic ```OSSL_HTTP_sendreq_new``` (or whatever the name should be). Things where were missing before, remain missing.

This PR just renames functions as described/evolved in the comment thread in the other PR, especially https://github.com/openssl/openssl/pull/13671#issuecomment-744054210

DDvO commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -4,112 +4,116 @@
-OCSP_set_max_response_length,
-OCSP_REQ_CTX_add1_header,
 OCSP_sendreq_bio,
+OSSL_HTTP_REQ_CTX_free,
+OSSL_HTTP_REQ_CTX_add1_header,
+OSSL_HTTP_REQ_CTX_set_max_response_length,
+OSSL_HTTP_REQ_CTX_i2d,
 OCSP_REQ_CTX_i2d,
+OCSP_REQ_CTX_add1_header,
+OCSP_REQ_CTX_free,
+OCSP_set_max_response_length,
 OCSP_REQ_CTX_set1_req
```
Please use consistent order of the functions here, in the synopsis, and in the description.

richsalz replied:
> Please use consistent order of the functions here, in the synopsis, and in the description.

done.

DDvO commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -126,7 +130,19 @@ L<OCSP_response_status(3)>
+OCSP_REQ_CTX_add1_header(),
+OCSP_REQ_CTX_set1_req(),
+OCSP_set_max_response_length()
+functions were deprecated in OpenSSL 3.0. OCSP_REQ_CTX_set1_req(rctx, req)
```
Please use a new line after the end of the sentence.

DDvO commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -4,112 +4,116 @@
 responder or B<NULL> if an error occurred.
 
-OCSP_REQ_CTX_free() and OCSP_set_max_response_length()
+OSSL_HTTP_REQ_CTX_fre() and OSSL_HTTP_REQ_CTX_set_max_response_length()
```
fre -> free

richsalz replied:
> fre -> free

done.

DDvO commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -126,7 +130,19 @@ L<OCSP_response_status(3)>
+
+The B<OCSP_REQ_CTX> typedef is an alias for B<OSSL_HTTP_REQ_CTX>.
+
+The OCSP_REQ_CTX_set1_req(), OCSP_REQ_CTX_free(),
+OCSP_REQ_CTX_add1_header(),
+OCSP_REQ_CTX_set1_req(),
+OCSP_set_max_response_length()
```
In this deprecation list I am missing
OCSP_REQ_CTX_new
OCSP_REQ_CTX_http
OCSP_REQ_CTX_i2d
OCSP_REQ_CTX_nbio
OCSP_REQ_CTX_nbio_d2i
OCSP_REQ_CTX_get0_mem_bio

richsalz replied:
> In this deprecation list I am missing

I added those that were documented; I did not mark undocumented functions as deprecated. I also put them in the same ordering as the fixed synopsis/text sections.

DDvO replied:
> I added those that were documented; I did not mark undocumented functions as deprecated.

All right, but `OCSP_REQ_CTX_i2d` is documented.

> I also put them in the same ordering as the fixed synopsis/text sections.

Looks like you forgot to push your changes.

richsalz replied:
> Looks like you forgot to push your changes.

i blame Santa.

DDvO reviewed (changes requested):
Thanks for taking care of this.
A couple of things to improve.

DDvO commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -53,6 +53,11 @@ L<openssl_user_macros(7)>:
 
 =head1 DESCRIPTION
 
+These functions only perform a minimal HTTP query to a responder.  For
+example, they only support HTTP POST queries to OCSP responders.  If an
+application needs more advanced features it should use a more complete
+HTTP library.
```
I think that meanwhile the sentence 
"If an application needs more advanced features it should use a more complete HTTP library." 
is not to-the-point anymore. The HTTP library has become more capable; in particular it does support GET queries. So I suggest dropping that sentence.

The two sentences before are still valid, though. 
I just ask to place each of them in a separate line since I learned some time ago that in POD files there should not be end of sentences in the middle of a line.

DDvO commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -66,6 +71,12 @@ should be set to B<"/ocspreq">
 
 OCSP_sendreq_nbio() performs I/O on the OCSP request context B<rctx>.
 When the operation is complete it returns the response in B<*presp>.
+It does not support setting a timeout or adding headers and is retained
+for  compatibility; use B<OCSP_sendreq_nbio()> instead.
```
This note does not belong here, but to `OCSP_sendreq_bio()`!
Please also remove the double space after "foo".

DDvO commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -134,12 +129,14 @@ The OCSP-specific names for types and functions were deprecated in OpenSSL 3.0.
-is equivalent to the following:
+OCSP_REQ_CTX_free(),
+OCSP_set_max_response_length(),
+and OCSP_REQ_CTX_set1_req(),
```
Please make also their order consistent with the above.
Nit: remove trailing `,`

richsalz replied:
The order does match the NAME and SYNOPSIS sections so I don't know what you want.  Removed the trailing comma.

DDvO replied:
Ah, right, the order is already consistent. Maybe Santa confused me :wink:

Yet i just noticed one last thing:
```
OCSP_REQ_CTX_set1_req(rctx, req) is equivalent to the following:

 OSSL_HTTP_REQ_CTX_i2d(rctx, "application/ocsp-request",
                       ASN1_ITEM_rptr(OCSP_REQUEST), (ASN1_VALUE *)req)
```
would fit better at the end of the DESCRIPTION section, 
mainly because otherwise `OCSP_REQ_CTX_set1_req()` is not mentioned there at all.

Once this is done, I think you can squash all commits since at least in my view it will be finished.

richsalz replied:
> Once this is done, I think you can squash all commits since at least in my view it will be finished.

Moved the text and squashed.  Thanks for the review!

richsalz commented:
fixup commit pushed.

DDvO reviewed (approved):
LGTM

mattcaswell commented on util/missingcrypto.txt:
```diff
@@ -803,11 +808,11 @@ OCSP_REQUEST_get_ext_by_critical(3)
+OSSL_HTTP_REQ_CTX_header(3)
+OSSL_HTTP_REQ_CTX_nbio(3)
+OSSL_HTTP_REQ_CTX_sendreq_d2i(3)
+OSSL_HTTP_REQ_CTX_new(3)
```
I think we need to document these things before this PR goes in (or document them in this PR). We are introducing a new namespace (OSSL_HTTP) here. We should not instantly pollute it with undocumented functions. I realise that these were undocumented before -  but this increases the size of the problem. We used to have one set of functions that were undocumented - we now have a new set of functions plus a set of aliases which are all undocumented.

richsalz replied:
>we need to document these things.

That's a reasonable attitude, but I am not going to do that. I think those involved with the original work should do so; they are free to take over this PR, or push commits, mail me files, whatever works.  (If I were paranoid, I would say that not wanting to do documentation was a motivation for the original define hacks in ```http_local.h``` :)

DDvO replied:
> That's a reasonable attitude, but I am not going to do that. I think those involved with the original work should do so; they are free to take over this PR, or push commits, mail me files, whatever works.

I second this.

> (If I were paranoid, I would say that not wanting to do documentation was a motivation for the original define hacks in http_local.h :)

My motivation for those "`#define` hacks" was a different one: I did not dare to touch the pretty old `OCSP_` function entries in `libcrypto.num`, fearing that doing so might break binary compatibility of `libcrypto` with older versions.

richsalz replied:
So @ddvo, it seems you wrote these functions (or someone at your company).  If that's correct, do you have any plans to get them documented?  And if I'm wrong, sorry, can you help figure out who did add them?

DDvO replied:
I did not write these functions - they have been around for many years.

I'm pretty sure that all of them were added by Stephen Henson - see for instance
```
git blame 29f178bddfdbd11218fbcba0b8060297696968e3^ crypto/ocsp/ocsp_ht.c|grep OCSP_REQ_CTX_new
6f9076ff370 (Dr. Stephen Henson      2012-11-28 16:22:53 +0000  64) OCSP_REQ_CTX *OCSP_REQ_CTX_new(BIO *io, int maxline)
```

According to `git blame`, the only other person having significant impact on that source file is Matt.
So @mattcaswell you seem the natural candidate to fill in the documentation for those old functions.

mattcaswell replied:
I think my "significant impact" amounts to having run the automated script "openssl-format-source" script over the entire codebase. So, by that measure, I think its my job to document *everything* :-)

DDvO replied:
BTW, I still wonder why Stephen Henson left the project - when he did I could not find any information on that.

mspncp replied:
> I think my "significant impact" amounts to having run the automated script "openssl-format-source" script over the entire codebase. So, by that measure, I think its my job to document everything :-)

Matt's commit 0f113f3ee4d is the great annotation wall. It's impossible to look through it, you need to climb over it if you want to blame other people.

richsalz replied:
Since @levitte wrote the missing docs, I am resolving this issue.

DDvO replied:
> Matt's commit 0f113f3 is the great annotation wall. It's impossible to look through it, you need to climb over it if you want to blame other people.

There is usually a way over, around, below, or through almost any wall :wink:
As far as analyzing the change history is concerned, as already mentioned above, there is a very simple way:
```
git blame 0f113f3^ crypto/ocsp/ocsp_ht.c
```

I'm unresolving this (part of the) issue just to make my comment to @mspncp better visible.

mattcaswell reviewed (changes requested):
See comment below. I also think this should have a CHANGES entry which outlines the change and says what people should do to change their code to avoid deprecation warnings.

levitte commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -4,117 +4,120 @@
+OSSL_HTTP_REQ_CTX_free,
+OSSL_HTTP_REQ_CTX_set_max_response_length,
+OSSL_HTTP_REQ_CTX_add1_header,
+OSSL_HTTP_REQ_CTX_i2d,
```
In this file, add a small note in the description saying that OCSP_REQ_CTX_i2d, OCSP_REQ_CTX_add1_header, OCSP_REQ_CTX_free, OCSP_set_max_response_length are replaced, and link to the new names.

richsalz replied:
> In this file, add a small note

Do you mean the new file (that someone else will be creating) or do you mean in `OCSP_sendreq_new` in this PR?

levitte replied:
With "this file", I mean `doc/man3/OCSP_sendreq_new.pod`, yes

richsalz replied:
I added a short note about deprecated replacements at the end of the DESCRIPTION section.

levitte commented on include/openssl/ocsp.h.in:
```diff
@@ -61,23 +61,26 @@ use OpenSSL::stackhash qw(generate_stack_macros);
+                                          const ASN1_ITEM *it);
+BIO *OSSL_HTTP_REQ_CTX_get0_mem_bio(OSSL_HTTP_REQ_CTX *rctx);
+void OSSL_HTTP_REQ_CTX_set_max_response_length(OSSL_HTTP_REQ_CTX *rctx,
+                                               unsigned long len);
```
The OSSL_HTTP_REQ_CTX functions should be declared in `include/openssl/http.h`

richsalz replied:
Moved the `OSSL_HTTP_REQ` functions to `http.h` as requested.

richsalz commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -126,7 +129,17 @@ L<OCSP_response_status(3)>
 =head1 HISTORY
 
-The OCSP_REQ_CTX_set1_req() function was deprecated in OpenSSL 3.0.
+The OCSP-specific names for types and functions were deprecated in OpenSSL 3.0.
```
I changed "The" to "Some"

levitte replied:
Since you enumerate them all a couple of lines down anyway, you can remove this line entirely

richsalz replied:
yes, just removing the line is better.  updated commit pushed.

richsalz commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -126,7 +129,17 @@ L<OCSP_response_status(3)>
-The OCSP_REQ_CTX_set1_req() function was deprecated in OpenSSL 3.0.
+The OCSP-specific names for types and functions were deprecated in OpenSSL 3.0.
+
+The B<OCSP_REQ_CTX> typedef is an alias for B<OSSL_HTTP_REQ_CTX>.
```
removed this line as requested.

richsalz commented:
Rebased to get the 2021 updates for generated files.  Added commit to address @levitte's feedback.

levitte commented:
> Added commit to address @levitte's feedback.

I assume there's going to be another update that includes documentation of OSSL_HTTP_REQ_CTX to `doc/man3/OSSL_HTTP_REQ_CTX.pod`

richsalz commented:
> I assume there's going to be another update that includes moving documentation of OSSL_HTTP_REQ_CTX to doc/man3/OSSL_HTTP_REQ_CTX.pod

To make sure I get it right:  You want all `OSSL_HTTP_xxx` things moved to a new manpage?

levitte commented:
> To make sure I get it right:  You want all `OSSL_HTTP_xxx` things moved to a new manpage?

All `OSSL_HTTP_REQ_CTX` things.  The rest of the `OSSL_HTTP_` stuff is already in `doc/man3/OSSL_HTTP_transfer.pod`.  And yeah, I do think it's worth having `OSSL_HTTP_REQ_CTX` in a separate manpage

richsalz commented:
I added a CHANGES entry and split some off into the new `OSSL_HTTP_REQ_free.pod` file.
There is still the question about whether or not the "new" (renamed) missing functions need documentation in order for this to be accepted.  Please decide.

levitte commented on doc/man3/OSSL_HTTP_REQ_CTX_free.pod:
```diff
@@ -0,0 +1,70 @@
+OSSL_HTTP_REQ_CTX_add1_header() adds header B<name> with value B<value> to the
+context B<rctx>. It can be called more than once to add multiple headers.
+It B<MUST> be called before any calls to OCSP_sendreq_nbio(). The B<req>
+parameter in the initial to OCSP_sendreq_new() call MUST be set to B<NULL> if
```
`OCSP_sendreq_nbio()` should be `OSSL_HTTP_REQ_CTX_sendreq_d2i()`
`OCSP_sendreq_new()` should be `OSSL_HTTP_REQ_CTX_new()`

richsalz replied:
Looks like you fixed this, too, in your commit.

richsalz commented:
@levitte thanks for the suggestions, they all seem reasonable.  I've stopped working on this PR until the project decides it will be accepted (the issue @mattcaswell raised at https://github.com/openssl/openssl/pull/13742#discussion_r551387954).  If not, and someone steps forward to document those functions, they are of course free to use this work if it will help.

levitte commented:
So, uhm, someone blocks this PR for the lack of documentation, but no one is stepping up to write that documentation...  am I getting the situation correctly?

richsalz commented:
This is a good summary: https://github.com/openssl/openssl/pull/13742#discussion_r551387954  The only thing worth adding is that I agreed to do the deprecation only.

levitte commented:
I wasn't trying to point fingers at anyone in particular, just noted that the situation is a kinda catch-22...

richsalz commented:
> I wasn't trying to point fingers at anyone

Sure, and I was likewise just trying to give a summary to help folks decide what to do.

I'm sure this is *exactly* the kind of thing you want to think about just returning from your holidays. :)

levitte commented:
> I'm sure this is *exactly* the kind of thing you want to think about just returning from your holidays. :)

Believe you me, this is light-weight :wink: 
BTW, I'm about to push a fixup commit here, that adds the missing documentation.

levitte commented:
There.  I took the opportunity to complete stuff I'd asked for that weren't done yet.

richsalz commented on doc/man3/OSSL_HTTP_REQ_CTX_free.pod:
```diff
@@ -0,0 +1,70 @@
+
+=head1 DESCRIPTION
+
+OSSL_HTTP_REQ_CTX_free() frees up the OCSP context B<rctx>.
```
I assume you fixed this (and other feedback about this file) in the commit you pushed; let me know if not.

richsalz commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -4,117 +4,89 @@
+                              const char *name, const char *value);
+ void OCSP_REQ_CTX_free(OCSP_REQ_CTX *rctx);
+ void OCSP_set_max_response_length(OSSL_HTTP_REQ_CTX *rctx,
+                                   unsigned long len);
```
I changed to using `OCSP_REQ_CTX` in the deprecated part of the SYNOPSIS section.

richsalz commented:
I also pushed a fixup commit that fixes a typo and also adds some additional warnings to the new manpage.

levitte commented:
So, uhm, @mattcaswell, does this meet your request at this point?

mattcaswell commented:
> So, uhm, @mattcaswell, does this meet your request at this point?

I've not re-reviewed this but, yes, it does. I note that the CHANGES entry seems too brief for my liking. We've got new preferred names for a whole load of functions and have introduced compatibility defines for them that are not available in a no-deprecated build. We should describe what people should do to fix their code.

levitte commented:
> I note that the CHANGES entry ...

I just noted that the OSSL_HTTP functions are previously not mentioned at all.  Not in CHANGES, not in NEWS.

levitte commented:
I just pushed an update of CHANGES that corresponds to this PR.

mattcaswell commented on crypto/ocsp/ocsp_http.c:
```diff
@@ -14,19 +14,20 @@
 
-OCSP_REQ_CTX *OCSP_sendreq_new(BIO *io, const char *path, OCSP_REQUEST *req,
-                               int maxline)
+OSSL_HTTP_REQ_CTX *OCSP_sendreq_new(BIO *io, const char *path,
```
This change in return type is also worthy of mention in the CHANGES entry IMO.

levitte replied:
The changes entry does say that the type OCSP_REQ_CTX has been replaced with OSSL_HTTP_REQ_CTX, is that not good enough?

mattcaswell replied:
So it does. Yeah ok.

mattcaswell commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -4,117 +4,89 @@
+If an application needs more advanced features it should use a more complete
+HTTP library.
+
+The function OCSP_sendreq_new() returns an B<OSSL_HTTP_REQ_CT> structure
```
Typo: `OSSL_HTTP_REQ_CTX`

levitte replied:
Fixed

mattcaswell commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -4,117 +4,89 @@
 default value of 4k is used. The OCSP request B<req> may be set to B<NULL>
 and provided later if required.
+The arguments to OCSP_sendreq_new() correspond to the components of the URL.
+For example if the responder URL is B<http://ocsp.com/ocspreq> the BIO
```
Perhaps "example.com" would be better?

levitte replied:
Sure

mattcaswell commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -4,117 +4,89 @@
-The function OCSP_sendreq_new() returns an B<OCSP_CTX> structure using the
+These functions only perform a minimal HTTP query to a responder.
+If an application needs more advanced features it should use a more complete
+HTTP library.
```
Should there be a reference to the OSSL_HTTP documentation somewhere here?

levitte replied:
Thanks, I'm re-editing this description

mattcaswell commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+read the response from (I<rbio>), the request method (I<method_GET>, which may
+be 1 to indicate that the C<GET> method is to be used, or 0 to indicate that
+the C<POST> method is to be used), the maximum expected response length
+(I<max_resp_len>), a response timeout (I<timeout>), the expected MIME
```
What is timeout measured in? Seconds?

levitte replied:
Yes (it relies on `BIO_do_connect_retry()`, which is undocumented [hint hint nudge nudge]

richsalz replied:
>Yes (it relies on `BIO_do_connect_retry()`, which is undocumented [hint hint nudge nudge])

? `doc/man3/BIO_socket_wait.pod`

mattcaswell commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+
+OSSL_HTTP_REQ_CTX_add1_header() adds header I<name> with value I<value> to the
+context I<rctx>. It can be called more than once to add multiple headers.
+For example, to add a C<Host> header for C<ocsp.com> you would call:
```
example.com?

levitte replied:
Fixed

mattcaswell commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+
+OSSL_HTTP_REQ_CTX_i2d() finalizes the HTTP request context by adding the DER
+encoding of I<req>, using the ASN.1 template I<it> to do the encoding.  The
+HTTP headers C<Content-Length> is automatically filled out, and the HTTP header
```
s/headers/header/

levitte replied:
Fixed

mattcaswell commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+
+=head1 WARNINGS
+
+The server's response may be unexpected if the hostname that the used to
```
that was used?

levitte replied:
Yup

mattcaswell commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -4,117 +4,89 @@
 
- OCSP_REQ_CTX *OCSP_sendreq_new(BIO *io, const char *path,
+ OSSL_HTTP_REQ_CTX *OCSP_sendreq_new(BIO *io, const char *path,
                                 OCSP_REQUEST *req, int maxline);
```
Indentation?

levitte replied:
Fixed

richsalz commented:
@levitte, are you going to address the latest round of feedback or do you want me to do so?  I'm fine to let you take this over from here.

levitte commented:
> @levitte, are you going to address the latest round of feedback or do you want me to do so? I'm fine to let you take this over from here.

Heh, I'm on it.

levitte commented:
There's been a rebase where squashing ended up being partially done, and where the newer squash and fixup commits have incorrect back references.  Ah well, I'll take it on me to perform the merge, so I'll deal.

mattcaswell commented:
I've approved. I guess there should be an explicit @levitte approval too to cover those parts of this PR provided by @richsalz.

levitte commented:
> I've approved. I guess there should be an explicit @levitte approval too to cover those parts of this PR provided by @richsalz.

Done

levitte commented:
@DDvO, you might want to re-approve...  so much has been done since yours...

DDvO commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+OSSL_HTTP_REQ_CTX_i2d() finalizes the HTTP request context by adding the DER
+encoding of I<req>, using the ASN.1 template I<it> to do the encoding.  The
+HTTP header C<Content-Length> is automatically filled out, and the HTTP header
+C<Content-Type> is also added, with I<content_type> as its value.  All of this
```
`C<Content-Type> is added, if I<content_type> is not NULL, with the given string value.`

DDvO commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+encoding of I<req>, using the ASN.1 template I<it> to do the encoding.  The
+HTTP header C<Content-Length> is automatically filled out, and the HTTP header
+C<Content-Type> is also added, with I<content_type> as its value.  All of this
+ends up in the internal memory B<BIO>.  This requires that the request type be
+C<POST>, i.e. that I<method_GET> is 0 in the OSSL_HTTP_REQ_CTX_new() call.
```
Good point!
`OSSL_HTTP_REQ_CTX_content()` should thrown an error if `rtcx->method_GET` is nonzero; could you please add this.

DDvO replied:
All right, will do along with fixing #13873.

DDvO commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+ends up in the internal memory B<BIO>.  This requires that the request type be
+C<POST>, i.e. that I<method_GET> is 0 in the OSSL_HTTP_REQ_CTX_new() call.
+
+OSSL_HTTP_REQ_CTX_nbio() performs the exchange of request and response via HTTP,
```
s/performs the exchange of/attempts exchanging/

levitte replied:
Yup

DDvO commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+call.  When successful, the contents of the internal memory B<BIO> is replaced
+with the contents of the HTTP response, without the response headers.
+
+OSSL_HTTP_REQ_CTX_sendreq_d2i() calls OSSL_HTTP_REQ_CTX_nbio(), and DER decodes
+the received response using the ASN.1 template I<it>.
```
```
OSSL_HTTP_REQ_CTX_sendreq_d2i() calls OSSL_HTTP_REQ_CTX_nbio() 
possibly several times until a timeout is reached, 
and DER decodes any received response using the ASN.1 template I<it>.
```

levitte replied:
Yup

DDvO commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+OSSL_HTTP_REQ_CTX_nbio() performs the exchange of request and response via HTTP,
+using the I<rbio> and I<wbio> that were given in the OSSL_HTTP_REQ_CTX_new()
+call.  When successful, the contents of the internal memory B<BIO> is replaced
+with the contents of the HTTP response, without the response headers.
```
add thereafter:
It may need to be called again if its result is -1, which indicates `BIO_should_retry`.
In such a case it is advisable to sleep a little in between to prevent a busy loop.

levitte replied:
Yup

DDvO commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+for I<rctx> to I<len>. If the response exceeds this length an error occurs.
+If not set a default value of 100k is used.
+
+OSSL_HTTP_REQ_CTX_get0_mem_bio() returns the internal memory B<BIO>.  This can
+be used to affect the HTTP request text.  I<Use with caution!>
```
BTW, as I learned, in POD files better place each sentence on a separate line.

levitte replied:
This was disputed against at some point, and to that, I shrugged with the response thaat as we don't have the endless dispute about number of spaces after a period, I'm OK with either.

DDvO replied:
OK by me, though that having each sentence on a separate line makes the file at least better maintainable (less brittle on small changes, and thus more easy to review those).

levitte replied:
Ah well, I consider this a discussion and change for another time

DDvO commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+=back
+
+When the request context is fully prepared, the HTTP exchange may be performed
+with OSSL_HTTP_REQ_CTX_nbio().
```
add: or OSSL_HTTP_REQ_CTX_sendreq_d2i().

levitte replied:
Thanks, added

DDvO reviewed (changes requested):
A couple of small additions to be done in `doc/man3/OCSP_sendreq_new.pod`
and several more in `doc/man3/OSSL_HTTP_REQ_CTX.pod`.
Thanks for documenting both the old and the new functions!

DDvO commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+
+=head1 SEE ALSO
+
+L<crypto(7)>
```
s/crypto(7)/OSSL_HTTP_transfer(3)/

levitte replied:
Fixed

DDvO commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+
+=head1 DESCRIPTION
+
+B<OSSL_HTTP_REQ_CTX> is a context structure for an HTTP request, used to
+collect all the necessary data to perform that request.
```
I suggest adding after this the following paragraph:

This file documents low-level HTTP functions rarely used directly.
High-level HTTP client functions like L<OSSL_HTTP_get(3)> are documented in L<OSSL_HTTP_transfer(3)>.

levitte replied:
Oh wow, this is the first time I actually look at the "higher level" functions, and that makes me wonder, why do we even have `OSSL_HTTP_REQ_CTX` publically at all?  Does that serve a purpose at all?

Mind you, yes, there are the `OCSP_REQ_CTX` functions, sure, but we don't *have* to make a 1:1 replacement, we only need to replace it with equivalent enough functionality, which the "higher level" OSSL_HTTP functions seem to be...

levitte replied:
... except of course, since we still need the macros to map `OSCP_REQ_CTX` functions, there must be something to map them to.  I'm quite tempted to propose that `OSSL_HTTP_REQ_CTX` be deprecated as well, which is a bit on the ridiculous side...

levitte replied:
Anyway, I added those lines nonetheless

mattcaswell replied:
> why do we even have OSSL_HTTP_REQ_CTX publically at all? Does that serve a purpose at all?

Please see #13873 (keeping in mind that `OSSL_HTTP_REQ_CTX` == `OCSP_REQ_CTX`

DDvO replied:
@levitte pleased to hear that you appreciate the new higher-level API.
As far as that goes, having `OSSL_HTTP_REQ_CTX` public only serves backward compatibility,
but when adding support for session keep-alive (i.e., multiple round trips per session) the context need to be handed out to the caller.

DDvO reviewed (changes requested):
A few further comments

levitte commented:
@DDvO's change request is big enough that there will be a need for re-approval.  Stay tuned

levitte commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -4,116 +4,87 @@
-OCSP_set_max_response_length() sets the maximum response length
-for B<rctx> to B<len>. If the response exceeds this length an error occurs.
-If not set a default value of 100k is used.
+The arguments to OCSP_sendreq_new() correspond to the components of the URL.
+For example if the responder URL is C<http://example.com/ocspreq> the BIO
+I<io> should be connected to host C<example.com> on port 80 and I<path>
+should be set to C</ocspreq>.
```
Fixed

levitte commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -4,116 +4,87 @@
-OCSP_sendreq_bio() performs an OCSP request using the responder B<io>, the URL
-path B<path>, the OCSP request B<req> and with a response header maximum line
+OCSP_sendreq_bio() is the same as a call to OCSP_sendreq_new() followed by
+OCSP_sendreq_nbio(), in a single call, with a response header maximum line
```
Added

levitte commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -4,116 +4,87 @@
+ OSSL_HTTP_REQ_CTX_i2d(rctx, "application/ocsp-request",
+                       ASN1_ITEM_rptr(OCSP_REQUEST), (ASN1_VALUE *)req)
+
+The other deprecated functions have been superseded by the following
+equivalent functions:
```
Added in both places

levitte commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -4,116 +4,87 @@
-parameter in the initial to OCSP_sendreq_new() call MUST be set to B<NULL> if
-additional headers are set.
+OCSP_sendreq_nbio() performs I/O on the OCSP request context I<rctx>.
+When the operation is complete it returns the response in I<*presp>.
```
I've rephrased that accordingly.

levitte commented on doc/man3/OCSP_sendreq_new.pod:
```diff
@@ -126,11 +97,17 @@ L<OCSP_response_status(3)>
 =head1 HISTORY
 
-The OCSP_REQ_CTX_set1_req() function was deprecated in OpenSSL 3.0.
+The
+OCSP_REQ_CTX_i2d(),
+OCSP_REQ_CTX_add1_header(),
+OCSP_REQ_CTX_free(),
+OCSP_set_max_response_length(),
+and OCSP_REQ_CTX_set1_req()
+functions were deprecated in OpenSSL 3.0.
```
Ah, that was another previously undocumented function...  thanks for pointing it out.

levitte replied:
Actually, considering exactly that, and there are a few more, I'm backing off adding new functions to this man-page.

DDvO replied:
My point was a different one: `OSSL_HTTP_REQ_CTX_get0_mem_bio()` is being documented in this file,
but the (only) thing still missing is to include it in the list of deprecated functions given here.

levitte replied:
Except that, no, it's not mentioned at all in this file... and it isn't in 1.1.1 either.

DDvO replied:
Ah, right, I confused `OCSP_REQ_CTX_get0_mem_bio` and `OSSL_HTTP_REQ_CTX_get0_mem_bio` here.
So no need to mention in this doc file that `OCSP_REQ_CTX_get0_mem_bio` has been deprecated.

levitte commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+gets populated with the B<BIO> to send the request to (I<wbio>), the B<BIO> to
+read the response from (I<rbio>), the request method (I<method_GET>, which may
+be 1 to indicate that the C<GET> method is to be used, or 0 to indicate that
+the C<POST> method is to be used), the maximum expected response length
```
Done, a little bit rephrased

levitte commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+read the response from (I<rbio>), the request method (I<method_GET>, which may
+be 1 to indicate that the C<GET> method is to be used, or 0 to indicate that
+the C<POST> method is to be used), the maximum expected response length
+(I<max_resp_len>), a response timeout measure in seconds (I<timeout>), the
```
Done

levitte commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+be 1 to indicate that the C<GET> method is to be used, or 0 to indicate that
+the C<POST> method is to be used), the maximum expected response length
+(I<max_resp_len>), a response timeout measure in seconds (I<timeout>), the
+expected MIME content type of the response (I<expected_content_type>), and a
```
Done

levitte commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+
+OSSL_HTTP_REQ_CTX_new() allocates a new HTTP request context structure, which
+gets populated with the B<BIO> to send the request to (I<wbio>), the B<BIO> to
+read the response from (I<rbio>), the request method (I<method_GET>, which may
```
Done

levitte commented on doc/man3/OSSL_HTTP_REQ_CTX.pod:
```diff
@@ -0,0 +1,172 @@
+structure (I<expect_asn1>).
+The allocated context structure is also populated with an internal allocated
+memory B<BIO>, which collects the HTTP request and additional headers as text.
+The returned context should only be used for a single HTTP request/response.
```
I assume you'll change the docs then, and take this as a side note

DDvO replied:
Yes.

levitte replied:
I'm sure that someone will give you all the memory support you might need :wink:

levitte commented:
My approval on stuff I haven't touched remains.  @mattcaswell, @DDvO, please re-review

DDvO commented:
@levitte, LGTM apart from a single leftover: list `SSL_HTTP_REQ_CTX_get0_mem_bio()` as depreated; see above.

DDvO commented:
Please squash as far as appropriate
and add to the commit extending `http.h`: `Fixes #12234`.

levitte commented:
> Please squash as far as appropriate
> and add to the commit extending `http.h`: `Fixes #12234`.

Done

DDvO commented:
@levitte, I'm not sure if you received a notification, but this should be ready to merge now.
Do you wanna do it? Else I can do the merge.

levitte commented:
I'm merging now

levitte commented:
Merged

83b6dc8dc739ce7ca82652461bea92c31e634a57 Deprecate OCSP_xxx API for OSSL_HTTP_xxx
